packages/web-api-scan-runner/src/runner/website-scan-result-updater.ts

```diff
@@ -0,0 +1,146 @@
+    }
+
+    private async executeUpdateWithRetries(action: () => Promise<void>, errorMessage: string): Promise<void> {
+        await this.retryHelper.executeWithRetries(
+            action,
+            async (error: Error) => {
+                this.logger.logError(errorMessage, {
+                    error: System.serializeError(error),
+                });
+            },
+            this.maxCombinedResultProcessingRetryCount,
+            1000,
+        );
+    }
+
+    private async generateCombinedScanResultsImpl(axeScanResults: AxeScanResults, pageScanResult: OnDemandPageScanResult): Promise<void> {
+        const websiteScanRef = this.getWebsiteScanRef(pageScanResult, 'consolidated-scan-report');
+        if (!websiteScanRef) {
+            return;
+        }
+
+        const websiteScanResult = await this.websiteScanResultProvider.read(websiteScanRef.id);
+        const combinedResultsBlobInfo = await this.combinedResultsBlobGetter.getBlobInfo(websiteScanResult.combinedResultsBlobId);
+        const combinedResultsBlobId = combinedResultsBlobInfo.blobId;
+
+        this.logger.setCommonProperties({
+            combinedResultsBlobId,
+            websiteScanId: websiteScanRef.id,
+        });
+
+        const combinedAxeResults = await this.axeResultMerger.mergeAxeResults(axeScanResults.results, combinedResultsBlobInfo);
+        const report = await this.reportSaver.save(
+            this.combinedReportGenerator.generate(
+                combinedAxeResults,
+                websiteScanResult,
+                axeScanResults.userAgent,
+                axeScanResults.browserResolution,
+            ),
+        );
+
+        const updatedWebsiteScanResults = {
+            id: websiteScanResult.id,
+            combinedResultsBlobId: combinedResultsBlobId,
+            reports: [report],
+            _etag: websiteScanResult._etag,
+        } as Partial<WebsiteScanResult>;
+
+        await this.updateWebsiteScanResult(updatedWebsiteScanResults, ' with combined result metadata');
+
+        if (report) {
+            pageScanResult.reports.push(report);
+        }
+    }
+
+    private async updateWebsiteScanResultWithDiscoveredUrlsImpl(
+        pageScanResult: OnDemandPageScanResult,
+        newlyDiscoveredUrls: string[],
+    ): Promise<void> {
+        const websiteScanRef = this.getWebsiteScanRef(pageScanResult, 'deep-scan');
+        if (!websiteScanRef) {
+            return;
+        }
+
+        const websiteScanResult = await this.websiteScanResultProvider.read(websiteScanRef.id);
+        const knownPages = websiteScanResult.knownPages;
+        const updatedKnownPages = this.urlDeduplicator.dedupe(knownPages, newlyDiscoveredUrls);
+
+        const updatedWebsiteScanResults: Partial<WebsiteScanResult> = {
+            id: websiteScanResult.id,
+            knownPages: updatedKnownPages,
+            _etag: websiteScanResult._etag,
+        };
+
+        return this.updateWebsiteScanResult(updatedWebsiteScanResults);
```

Comment:
It _may_ be better to use `await` here rather then `return`. My understanding is that using await doesn't add much overhead and makes it possible to get a meaningful callstack when debugging.